packages/taquito-sapling/src/types.ts

```diff
@@ -23,11 +36,129 @@ export interface Ciphertext {
   nonceOut: Buffer;
 }
 
-export interface SaplingTransaction {
-  spendDescriptions: SpendDescription[];
-  outputDescriptions: OutputDescription[];
+export interface SaplingTransactionInput {
+  cv: Buffer;
+  nf: Buffer;
+  rk: Buffer;
+  proof: Buffer;
   signature: Buffer;
+}
+
+export interface SaplingSpendDescription extends SaplingTransactionInput {
+  rt: Buffer;
+}
+
+export interface SaplingTransactionOutput {
+  cm: Buffer;
+  proof: Buffer;
+  ciphertext: Ciphertext;
+}
+
+export interface ParametersSaplingTransaction {
+  to: string;
+  amount: number;
+  memo?: string;
+  mutez?: boolean;
+}
+
+export type ParametersUnshieldedTransaction = Omit<ParametersSaplingTransaction, 'memo'>;
+
+export interface ParametersSpendProof {
+  saplingContext: number;
+  sk: Uint8Array;
+  address: Uint8Array;
+  rcm: Uint8Array;
+  ar: Buffer;
+  amount: string;
+  root: string;
+  witness: string;
+}
+
+export interface ParametersSpendSig {
+  sk: Uint8Array;
+  ar: Buffer;
+  unsignedSpendDescription: Omit<SaplingSpendDescription, 'signature'>;
+  hash: Uint8Array;
+}
+
+export interface ParametersOutputDescription {
+  saplingContext: number;
+  address: Uint8Array;
+  amount: string;
+  memo: string;
+  rcm: Buffer;
+  ovk?: Buffer;
+}
+
+export interface ParametersOutputProof {
+  saplingContext: number;
+  esk: Buffer;
+  address: Uint8Array;
+  rcm: Uint8Array;
+  amount: string;
+}
+
+export interface ParametersBindingSig {
+  saplingContext: number;
+  inputs: SaplingTransactionInput[];
+  outputs: SaplingTransactionOutput[];
   balance: BigNumber;
-  root: Buffer;
   boundData: Buffer;
 }
+
+export interface ParametersCiphertext {
+  address: Uint8Array;
+  esk: Buffer;
+  diversifier: Buffer;
+  ock: Uint8Array;
+  amount: string;
+  rcm: Buffer;
+  memo: string;
+}
+
+export interface SaplingTransactionPlaintext {
+  diversifier: Buffer;
+  amount: string;
+  rcm: Buffer;
+  memoSize: number;
+  memo: string;
+}
+
+export interface Input {
+  value: Uint8Array;
+  memo: Uint8Array;
+  paymentAddress: Uint8Array;
+  rcm: Uint8Array;
+  position: number;
+}
+
+export type SaplingContractId =
+  | {
+      contractAddress?: never;
+      saplingId: string;
+    }
+  | {
+      contractAddress: string;
+      saplingId?: never;
+    };
```

Comment:
not sure if i understand how this works as a type 
is it the same as `{contractAddress: string} | {saplingId: string}`?

Reply:
Yes it is the same

Reply:
would it be more clear just to have it without the other option even with the `never`?

Reply:
Might not be totally equivalent fianlly. The change caused a build issue:
```
src/sapling-tx-viewer/sapling-transaction-viewer.ts(120,33): error TS2339: Property 'saplingId' does not exist on type 'SaplingContractId'.
  Property 'saplingId' does not exist on type '{ contractAddress: string; }'.
src/sapling-tx-viewer/sapling-transaction-viewer.ts(122,39): error TS2339: Property 'saplingId' does not exist on type 'SaplingContractId'.
  Property 'saplingId' does not exist on type '{ contractAddress: string; }'.
src/sapling-tx-viewer/sapling-transaction-viewer.ts(125,40): error TS2339: Property 'contractAddress' does not exist on type 'SaplingContractId'.
  Property 'contractAddress' does not exist on type '{ saplingId: string; }'.
src/sapling-tx-viewer/sapling-transaction-viewer.ts(127,33): error TS2339: Property 'contractAddress' does not exist on type 'SaplingContractId'.
  Property 'contractAddress' does not exist on type '{ saplingId: string; }'.
```